File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        countT={}
        for i in t:
            countT[i]=1+countT.get(i,0)
        window={}
        have=0
        need=len(countT)
        res=[-1,-1]
        resLen=float(inf)
        l=0
        for r in range(len(s)):
            c = s[r]
            window[c] = 1+ window.get(c,0)
            if c in countT and window[c]==countT[c]:
                have+=1
            while(have==need):
                if (r-l+1) < resLen:
                    resLen=r-l+1
                    res[0]=l
                    res[1]=r

                window[s[l]]-=1
                if s[l] in countT and window[s[l]]<countT[s[l]]:
                    have-=1
                l+=1
        return s[res[0]:res[1]+1]     

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        // a map for t freq
        HashMap<Character, Integer> tmap = new HashMap<>();
        for(char c : t.toCharArray()){
            tmap.put(c,tmap.getOrDefault(c,0)+1);
        }

        
        int st=0;
        int have = 0;
        int need = tmap.size();
        int min_len = Integer.MAX_VALUE;
        int start = 0;
        // keeping a track of window with map 
        HashMap<Character, Integer> smap = new HashMap<>();
        for(int end = 0; end < s.length(); end++){
            char c = s.charAt(end);
            smap.put(c,smap.getOrDefault(c,0)+1);
            if(tmap.containsKey(c) && tmap.get(c).equals(smap.get(c))){
                have++;
            }
            while(have == need){
                // min_len = Math.min(min_len,end-st+1);
                char x = s.charAt(st);
                if((end-st+1) < min_len){
                    min_len = end-st+1;
                    start = st;
                }
                smap.put(x,smap.get(x)-1);
                // if(smap.get(s.charAt(st)) <= 0){
                //     smap.remove(s.charAt(st));
                // }
                
                if(tmap.containsKey(x) && (smap.get(x) < tmap.get(x))){
                    have--;
                }

            st++;
            }
        }
        if(min_len == Integer.MAX_VALUE) {return ""; }
        else{
        return s.substring(start,start+min_len); }
    }
}